runtime/compiler/net/RawTypeConvert.hpp

```diff
@@ -0,0 +1,290 @@
+      operator bool() const { return false; }
+      };
+
+   // The following are the template functions used to serialize/deserialize data.
+   // RawTypeConvert<T>::onRecv(Message::DataDescriptor *desc) - given a data descriptor,
+   // deserializes it into a value of type T and returns it.
+   //
+   // RawTypeConvert::onSend(Message &msg, const T&val) - given a value of type T,
+   // serializes it into an instance of Message::DataDescriptor and adds it to the message.
+   // Returns the number of data bytes written.
+   template <typename T, typename = void> struct RawTypeConvert { };
+   template <> struct RawTypeConvert<uint32_t>
+      {
+      static inline uint32_t onRecv( Message::DataDescriptor *desc) { return *static_cast<uint32_t *>(desc->getDataStart()); }
+      static inline uint32_t onSend(Message &msg, const uint32_t &val)
+         {
+         msg.addData(Message::DataDescriptor(Message::DataDescriptor::DataType::UINT32, sizeof(uint32_t)), &val);
+         return sizeof(uint32_t);
+         }
+      };
+   template <> struct RawTypeConvert<uint64_t>
+      {
+      static inline uint64_t onRecv( Message::DataDescriptor *desc) { return *static_cast<uint64_t *>(desc->getDataStart()); }
+      static inline uint32_t onSend(Message &msg, const uint64_t &val)
+         {
+         msg.addData(Message::DataDescriptor(Message::DataDescriptor::DataType::UINT64, sizeof(uint64_t)), &val);
+         return sizeof(uint64_t);
+         }
+      };
+   template <> struct RawTypeConvert<int32_t>
+      {
+      static inline int32_t onRecv( Message::DataDescriptor *desc) { return *static_cast<int32_t *>(desc->getDataStart()); }
+      static inline uint32_t onSend(Message &msg, const int32_t &val)
+         {
+         msg.addData(Message::DataDescriptor(Message::DataDescriptor::DataType::INT32, sizeof(int32_t)), &val);
+         return sizeof(int32_t);
+         }
+      };
+   template <> struct RawTypeConvert<int64_t>
+      {
+      static inline int64_t onRecv( Message::DataDescriptor *desc) { return *static_cast<int64_t *>(desc->getDataStart()); }
+      static inline uint32_t onSend(Message &msg, const int64_t &val)
+         {
+         msg.addData(Message::DataDescriptor(Message::DataDescriptor::DataType::INT64, sizeof(int64_t)), &val);
+         return sizeof(int64_t);
+         }
+      };
+   template <> struct RawTypeConvert<bool>
+      {
+      static inline bool onRecv( Message::DataDescriptor *desc) { return *static_cast<bool *>(desc->getDataStart()); }
+      static inline uint32_t onSend(Message &msg, const bool &val)
+         {
+         msg.addData(Message::DataDescriptor(Message::DataDescriptor::DataType::BOOL, sizeof(bool)), &val);
+         return sizeof(bool);
+         }
+      };
+   template <> struct RawTypeConvert<const std::string>
+      {
+      static inline std::string onRecv( Message::DataDescriptor *desc)
+         {
+         return std::string(static_cast<char *>(desc->getDataStart()), desc->size);
```

Comment:
Yes, it would. Not sure what the alternate solution would be.
I guess we could have a static/global string, copy the received string there and return a reference. That way, the number of memory allocations would be minimized.

Reply:
I posted a comment in the PR on how we can do this. Emulate std::span and use it to replace it the std::string. Again, this is an optimization for a follow on PR.